File: blue-core/src/main/java/blue/utility/CSDUtility.java
package blue.utility;

/**
 * Title: blue Description: an object composition environment for csound
 * Copyright: Copyright (c) 2001 Company: steven yi music
 *
 * @author steven yi
 * @version 1.0
 */
import blue.Arrangement;
import blue.BlueData;
import blue.BlueSystem;
import blue.SoundLayer;
import blue.orchestra.GenericInstrument;
import blue.soundObject.*;
import blue.time.TempoPoint;
import blue.time.TimeContext;
import blue.time.TimePosition;
import blue.udo.OpcodeList;
import blue.udo.UserDefinedOpcode;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.util.*;
import java.util.Map.Entry;

public class CSDUtility {

    public static final int IMPORT_GLOBAL = 0;

    public static final int IMPORT_SINGLE_SOUNDOBJECT = 1;

    public static final int IMPORT_SOUNDOBJECT_PER_INSTRUMENT = 2;

    public CSDUtility() {
    }

    public static BlueData convertOrcScoToBlue(File orcFile, File scoFile,
            int importMode) {

        BlueData data = null;

        try {
            String orc = TextUtilities.getTextFromFile(orcFile);
            String sco = TextUtilities.getTextFromFile(scoFile);

            data = new BlueData();

            parseCsOrc(data, orc);
            parseCsScore(data, sco, importMode);
        } catch (Exception e) {
            e.printStackTrace();
        }

        return data;
    }

    public static BlueData convertCSDtoBlue(File csdFile, int importMode) {
        BlueData data = null;

        try {
            String CSD = TextUtilities.getTextFromFile(csdFile);
            data = convertCSDtoBlue(CSD, importMode);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return data;
    }

    private static BlueData convertCSDtoBlue(String CSD, int importMode) {
        BlueData data = new BlueData();

        // String csOptions = TextUtilities.getTextBetweenTags("CsOptions",
        // CSD);
        // if (csOptions != null && csOptions.trim().length() > 0) {
        // data.getProjectProperties().CsOptions = csOptions.trim();
        // }
        String orc = TextUtilities.getTextBetweenTags("CsInstruments", CSD);
        String sco = TextUtilities.getTextBetweenTags("CsScore", CSD);

        if (orc != null) {
            parseCsOrc(data, orc);
        }
        if (sco != null) {
            parseCsScore(data, sco, importMode);
        }

        if (data.getScore().get(0).isEmpty()) {
            PolyObject pObj = (PolyObject) data.getScore().get(0);
            pObj.add(new SoundLayer());
        }

        return data;
    }

    protected static void parseCsScore(BlueData data, String scoreText, int importMode) {
        StringBuilder tables = new StringBuilder();
        StringBuilder iStatements = new StringBuilder();

        String[] lines = scoreText.split("\n");
        String line;

        for (int i = 0; i < lines.length; i++) {
            line = lines[i].trim();

            if (line.startsWith("f")) {
                tables.append(lines[i]).append("\n");

                if (i < lines.length - 1) {
                    do {
                        String nextLine = lines[i + 1].trim();

                        if (nextLine.length() == 0) {
                            break;
                        }

                        char c = nextLine.charAt(0);
                        if (Character.isDigit(c) || c == '\"' || c == '.') {
                            tables.append(lines[i + 1]).append("\n");
                            i++;
                        } else {
                            break;
                        }
                    } while (i < lines.length - 1);
                }

            } else if (line.startsWith("i")) {
                iStatements.append(line).append("\n");

                if (i < lines.length - 1) {
                    do {
                        String nextLine = lines[i + 1].trim();

                        if (nextLine.length() == 0) {
                            break;
                        }

                        char c = nextLine.charAt(0);
                        if (Character.isDigit(c) || c == '\"' || c == '.') {
                            iStatements.append(lines[i + 1]).append("\n");
                            i++;
                        } else {
                            break;
                        }
                    } while (i < lines.length - 1);
                }
            } else if (line.startsWith("s")) {
                iStatements.append(line).append("\n");
            } else if (line.startsWith("t")) {
                if (line.length() > 1) {
                    line = line.substring(1).trim();
                    var tempoMap = data.getScore().getTempoMap();
                    var parts = line.split("\\s+");

                    if (parts.length % 2 == 0) {
                        try {
                            tempoMap.reset();
                            for (int j = 0; j < parts.length; j += 2) {
                                double beat = Double.parseDouble(parts[j]);
                                double tempo = Double.parseDouble(parts[j + 1]);
                                if (j == 0) {
                                    // First point - set it (reset already created one at beat 0)
                                    tempoMap.setTempoPoint(0, beat, tempo);
                                } else {
                                    tempoMap.addTempoPoint(new TempoPoint(beat, tempo));
                                }
                            }
                            tempoMap.setEnabled(true);
                        } catch (Exception e) {
                            throw new RuntimeException("Invalid tempo statement found");
                        }
                    } else {
                        throw new RuntimeException("Invalid tempo statement found");
                    }
                }

            }

        }

        String noteText = iStatements.toString();
        ScoreSection[] sections;

        switch (importMode) {
            case IMPORT_GLOBAL:
                data.getScore().get(0).newLayerAt(-1);
                data.getGlobalOrcSco().setGlobalSco(noteText);
                break;

            case IMPORT_SINGLE_SOUNDOBJECT:
                sections = getScoreSections(noteText);
                for (ScoreSection scoreSection : sections) {
                    setSoundObjectPerSection(data, scoreSection);
                }
                break;

            case IMPORT_SOUNDOBJECT_PER_INSTRUMENT:
                sections = getScoreSections(noteText);
                for (ScoreSection section : sections) {
                    setSoundObjectsPerInstrument(data, section);
                }

                break;

        }

        data.getTableSet().setTables(tables.toString());
    }

    private static ScoreSection[] getScoreSections(String scoreText) {
        ArrayList<ScoreSection> scoreSections = new ArrayList<>();
        StringTokenizer st = new StringTokenizer(scoreText, "\n");
        String line = "";

        StringBuffer currentSection = new StringBuffer();
        double sectionStartTime = 0.0f;

        while (st.hasMoreTokens()) {
            line = st.nextToken().trim();

            if (line.startsWith("s")) {
                ScoreSection section = new ScoreSection();
                section.scoreText = currentSection.toString();
                section.sectionStartTime = sectionStartTime;

                scoreSections.add(section);

                NoteList nl;

                try {
                    nl = ScoreUtilities.getNotes(section.scoreText);
                } catch (NoteParseException e) {
                    throw new RuntimeException(e);
                }
                sectionStartTime += ScoreUtilities.getTotalDuration(nl);
                currentSection = new StringBuffer();

            } else {
                currentSection.append(line).append("\n");

            }
        }

        ScoreSection section = new ScoreSection();
        section.scoreText = currentSection.toString();
        section.sectionStartTime = sectionStartTime;

        scoreSections.add(section);

        ScoreSection[] returnSections = new ScoreSection[scoreSections.size()];

        for (int i = 0; i < scoreSections.size(); i++) {
            returnSections[i] = (ScoreSection) scoreSections.get(i);
        }

        return returnSections;
    }

    private static void setSoundObjectPerSection(BlueData data,
            ScoreSection section) {
        TimeContext context = data.getScore().getTimeContext();
        GenericScore genScore = createSizedGenericScore(section.scoreText,
                BlueSystem.getString("csd.importedScore"), context);

        genScore.setStartTime(TimePosition.beats(section.sectionStartTime));

        PolyObject pObj = (PolyObject) data.getScore().get(0);

        SoundLayer sLayer = pObj.newLayerAt(-1);
        sLayer.add(genScore);

    }

    // TODO - Possible problems with score processing here if SCO uses carry's,
    // ramp's, etc.
    private static void setSoundObjectsPerInstrument(BlueData data,
            ScoreSection section) {
        TimeContext context = data.getScore().getTimeContext();
        TreeMap<Integer, StringBuffer> map = new TreeMap<>();

        StringTokenizer st = new StringTokenizer(section.scoreText, "\n");
        String line = "";

        Note previousNote = null;
        Note note;
        Integer iNum;
        StringBuffer buffer;

        while (st.hasMoreTokens()) {
            line = st.nextToken();

            note = null;

            try {
                note = Note.createNote(line, previousNote);
            } catch (NoteParseException e) {
                // TODO Auto-generated catch block
                e.printStackTrace();
            }

            if (note == null) {
                continue;
            }

            iNum = Integer.parseInt(note.getPField(1));

            if (map.containsKey(iNum)) {
                buffer = (StringBuffer) map.get(iNum);
                buffer.append(line).append("\n");
            } else {
                buffer = new StringBuffer();
                buffer.append(line).append("\n");
                map.put(iNum, buffer);
            }
            previousNote = note;
        }

        SoundLayer sLayer;

        for (Entry<Integer, StringBuffer> entry : map.entrySet()) {
            iNum = entry.getKey();
            buffer = entry.getValue();

            if (buffer == null) {
                continue;
            }

            PolyObject pObj = (PolyObject) data.getScore().get(0);
            sLayer = pObj.newLayerAt(-1);

            String score = buffer.toString();
            NoteList notes;

            try {
                notes = ScoreUtilities.getNotes(score);
            } catch (NoteParseException e) {
                throw new RuntimeException(e);
            }

            notes.sort();
            double minStart = notes.get(0).getStartTime();

            ScoreUtilities.normalizeNoteList(notes);

            GenericScore genScore = createSizedGenericScore(notes.toString(),
                    "Instrument " + iNum.toString(), context);

            genScore.setStartTime(TimePosition.beats(minStart + section.sectionStartTime));

            sLayer.add(genScore);

        }
    }

    private static GenericScore createSizedGenericScore(String noteText,
            String name, TimeContext context) {

        GenericScore genScore = new GenericScore();
        genScore.setText(noteText);
        genScore.setSubjectiveDuration(genScore.getObjectiveDuration(context));
        genScore.setName(name);
        return genScore;
    }

    private static void parseCsOrc(BlueData data, String orc) {
        StringTokenizer st = new StringTokenizer(orc, "\n");

        String line = "";

        StringBuilder globalOrch = new StringBuilder();

        String sr = null;
        String kr = null;
        String ksmps = null;

        String instrIds = "";

        StringBuffer iBody = new StringBuffer();
        StringBuilder udoDeclaration = null;

        UserDefinedOpcode udo = null;
        GenericInstrument instr = null;

        Arrangement arrangement = data.getArrangement();
        OpcodeList opcodeList = data.getOpcodeList();

        int state = 0;

        boolean reprocessCurrentLine = false;

        while (reprocessCurrentLine || st.hasMoreTokens()) {
            if (!reprocessCurrentLine) {
                line = st.nextToken();
            } else {
                reprocessCurrentLine = false;
            }
            String trimLine = line.trim();

            switch (state) {
                case 0:
                    if (trimLine.startsWith("instr")) {
                        int index = line.indexOf(';');
                        String iName = "";

                        if (index != -1) {
                            iName = line.substring(index + 1).trim();
                            line = line.substring(0, index);
                        }
                        instrIds = line.substring(line.indexOf("instr") + 5)
                                .trim();

                        instr = new GenericInstrument();
                        instr.setName(iName);

                        state = 1;

                    } else if (trimLine.startsWith("opcode")) {
                        int index = line.indexOf(';');
                        if (index != -1) {
                            line = line.substring(0, index);
                        }
                        udo = UDOUtilities.parseUDODeclaration(line);

                        if (udo != null) {
                            state = 2;
                        } else {
                            udoDeclaration = new StringBuilder(line.trim());
                            state = 3;
                        }
                    } else {
                        if (trimLine.startsWith("kr")) {
                            kr = line.substring(line.indexOf('=') + 1).trim();
                        } else if (trimLine.startsWith("sr")) {
                            sr = line.substring(line.indexOf('=') + 1).trim();
                        } else if (trimLine.startsWith("nchnls")) {
                            data.getProjectProperties().channels = line
                                    .substring(line.indexOf('=') + 1).trim();
                        } else if (trimLine.startsWith("ksmps")) {
                            ksmps = line.substring(line.indexOf('=') + 1)
                                    .trim();
                        } else {
                            globalOrch.append(line).append("\n");
                        }
                    }
                    break;
                case 1:
                    if (trimLine.startsWith("endin")) {

                        if (instr != null && instrIds != null) {
                            instr.setText(iBody.toString());

                            if (instrIds.indexOf(',') > -1) {
                                String[] ids = instrIds.split(",");

                                for (String id : ids) {
                                    arrangement.insertInstrument(id, instr);
                                }

                            } else {
                                arrangement.insertInstrument(instrIds, instr);
                            }

                        }

                        instr = null;
                        instrIds = null;
                        iBody = new StringBuffer();

                        state = 0;
                    } else {
                        if (instr != null) {
                            iBody.append(line).append("\n");
                        }
                    }
                    break;
                case 2:
                    if (trimLine.startsWith("endop")) {
                        if (udo != null) {
                            UDOUtilities.finalizeParsedUDO(udo, iBody.toString());
                            opcodeList.addOpcode(udo);
                            iBody = new StringBuffer();

                            udo = null;
                        }
                        state = 0;
                    } else {
                        if (udo != null) {
                            iBody.append(line).append("\n");
                        }
                    }
                    break;
                case 3:
                    if (UDOUtilities.isInstrOrUDODeclarationBoundary(trimLine)) {
                        udoDeclaration = null;
                        state = 0;
                        reprocessCurrentLine = true;
                        break;
                    }

                    if (!trimLine.isBlank()) {
                        udoDeclaration.append("\n").append(trimLine);
                    }

                    udo = UDOUtilities.parseUDODeclaration(udoDeclaration.toString());
                    if (udo != null) {
                        udoDeclaration = null;
                        state = 2;
                    }
                    break;
            }
        }

        /* HANDLE RESERVED GLOBAL VARIABLES */
        if (kr != null && ksmps == null) {
            try {
                double krDouble = Double.parseDouble(kr);
                double srDouble = Double.parseDouble(sr);
                ksmps = Integer.toString((int) (srDouble / krDouble));
            } catch (NumberFormatException nfe) {
                ksmps = null;
            }
        }

        if (sr != null) {
            data.getProjectProperties().setSampleRate(sr);
        }

        if (ksmps != null) {
            data.getProjectProperties().ksmps = ksmps;
        }

        data.getGlobalOrcSco().setGlobalOrc(globalOrch.toString());
    }

    public static void main(String[] args) {
        File test = new File("/work/blue/trappedInConvert/01-Trapped.csd");

        try (BufferedReader in = new BufferedReader(new FileReader(test))) {
            String line = "";
            StringBuilder buffer = new StringBuilder();
            while ((line = in.readLine()) != null) {
                buffer.append(line).append("\n");
            }
            String CSD = buffer.toString();

            BlueData data = CSDUtility.convertCSDtoBlue(CSD,
                    CSDUtility.IMPORT_SINGLE_SOUNDOBJECT);

            System.out.println(data.saveAsXML());

        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }
    }
}

class ScoreSection {

    String scoreText = "";

    double sectionStartTime = 0.0f;
}

File: packages/blue-data/src/utilities/csd-utility.ts
/**
 * CSDUtility — imports CSD and ORC/SCO files into BlueData projects.
 * Mirrors the Java CSDUtility class.
 */
import { BlueData } from '../blue-data';
import { GenericInstrument } from '../instruments/generic-instrument';
import { GenericScore } from '../sound-objects/generic-score';
import { PolyObject } from '../sound-objects/poly-object';
import { Note } from '../sound-objects/note';
import { TimePosition } from '../time/time-position';
import { TimeDuration } from '../time/time-duration';
import { TempoPoint } from '../time/tempo-point';
import { CurveType } from '../time/curve-type';
import { TimeContext } from '../time/time-context';
import { parseUDODeclaration } from '../opcodes/udo-utilities';
import { OpcodeDefinition } from '../opcodes/opcode-definition';
import { getNotes, getTotalDuration } from './score';

export enum CSDImportMode {
  IMPORT_GLOBAL = 0,
  IMPORT_SINGLE_SOUNDOBJECT = 1,
  IMPORT_SOUNDOBJECT_PER_INSTRUMENT = 2,
}

interface ScoreSection {
  scoreText: string;
  sectionStartTime: number;
}

/**
 * Extract content between XML-style tags, such as <CsInstruments>...</CsInstruments>.
 * Mirrors Java TextUtilities.getTextBetweenTags().
 */
export function getTextBetweenTags(tagName: string, text: string): string | null {
  const startTag = `<${tagName}>`;
  const endTag = `</${tagName}>`;

  const startIdx = text.indexOf(startTag);
  if (startIdx === -1) return null;

  const contentStart = startIdx + startTag.length;
  const endIdx = text.indexOf(endTag, contentStart);
  if (endIdx === -1) return null;

  return text.substring(contentStart, endIdx);
}

export function convertOrcScoToBlue(
  orcText: string,
  scoText: string,
  importMode: CSDImportMode = CSDImportMode.IMPORT_GLOBAL,
): BlueData {
  const data = new BlueData();

  parseCsOrc(data, orcText);
  parseCsScore(data, scoText, importMode);

  const rootScore = data.getScore();
  if (rootScore.length > 0 && rootScore[0] instanceof PolyObject) {
    const pObj = rootScore[0] as PolyObject;
    if (pObj.length === 0) {
      pObj.newLayerAt(-1);
    }
  }

  return data;
}

export function convertCSDtoBlue(
  csdText: string,
  importMode: CSDImportMode = CSDImportMode.IMPORT_GLOBAL,
): BlueData {
  const data = new BlueData();

  const orc = getTextBetweenTags('CsInstruments', csdText);
  const sco = getTextBetweenTags('CsScore', csdText);

  if (orc !== null) {
    parseCsOrc(data, orc);
  }
  if (sco !== null) {
    parseCsScore(data, sco, importMode);
  }

  const rootScore = data.getScore();
  if (rootScore.length > 0 && rootScore[0] instanceof PolyObject) {
    const pObj = rootScore[0] as PolyObject;
    if (pObj.length === 0) {
      pObj.newLayerAt(-1);
    }
  }

  return data;
}

export function parseCsOrc(data: BlueData, orc: string): void {
  const lines = orc.split('\n');

  let globalOrch = '';
  let sr: string | null = null;
  let kr: string | null = null;
  let ksmps: string | null = null;

  let instrIds: string | null = null;
  let iBody = '';
  let udoDeclaration: string | null = null;

  let udo: OpcodeDefinition | null = null;
  let instr: GenericInstrument | null = null;

  const arrangement = data.getArrangement();
  const opcodeList = data.getOpcodeList();

  let state = 0;
  let reprocessCurrentLine = false;
  let i = 0;

  while (reprocessCurrentLine || i < lines.length) {
    let line: string;
    if (!reprocessCurrentLine) {
      line = lines[i];
    } else {
      line = lines[i];
      reprocessCurrentLine = false;
    }

    const trimLine = line.trim();

    switch (state) {
      case 0: {
        if (trimLine.startsWith('instr')) {
          const index = line.indexOf(';');
          let iName = '';

          if (index !== -1) {
            iName = line.substring(index + 1).trim();
            line = line.substring(0, index);
          }
          instrIds = line.substring(line.indexOf('instr') + 5).trim();

          instr = new GenericInstrument();
          instr.setName(iName);

          state = 1;
        } else if (trimLine.startsWith('opcode')) {
          let declLine = line;
          const index = declLine.indexOf(';');
          if (index !== -1) {
            declLine = declLine.substring(0, index);
          }
          udo = parseUDODeclaration(declLine);

          if (udo != null) {
            state = 2;
          } else {
            udoDeclaration = declLine.trim();
            state = 3;
          }
        } else {
          if (trimLine.startsWith('kr')) {
            kr = line.substring(line.indexOf('=') + 1).trim();
          } else if (trimLine.startsWith('sr')) {
            sr = line.substring(line.indexOf('=') + 1).trim();
          } else if (trimLine.startsWith('nchnls')) {
            data.getProjectProperties().channels = line.substring(line.indexOf('=') + 1).trim();
          } else if (trimLine.startsWith('ksmps')) {
            ksmps = line.substring(line.indexOf('=') + 1).trim();
          } else {
            globalOrch += line + '\n';
          }
        }
        break;
      }
      case 1: {
        if (trimLine.startsWith('endin')) {
          if (instr != null && instrIds != null) {
            instr.setText(iBody);

            if (instrIds.indexOf(',') > -1) {
              const ids = instrIds.split(',');

              for (const id of ids) {
                arrangement.addInstrumentWithId(instr, id.trim());
              }
            } else {
              arrangement.addInstrumentWithId(instr, instrIds);
            }
          }

          instr = null;
          instrIds = null;
          iBody = '';
          state = 0;
        } else {
          if (instr != null) {
            iBody += line + '\n';
          }
        }
        break;
      }
      case 2: {
        if (trimLine.startsWith('endop')) {
          if (udo != null) {
            udo.setCode(iBody);
            opcodeList.addOpcode(udo);
            iBody = '';
            udo = null;
          }
          state = 0;
        } else {
          if (udo != null) {
            iBody += line + '\n';
          }
        }
        break;
      }
      case 3: {
        if (isInstrOrUDODeclarationBoundary(trimLine)) {
          udoDeclaration = null;
          state = 0;
          reprocessCurrentLine = true;
          break;
        }

        if (trimLine.length > 0) {
          udoDeclaration = (udoDeclaration ?? '') + '\n' + trimLine;
        }

        if (udoDeclaration != null) {
          udo = parseUDODeclaration(udoDeclaration);
          if (udo != null) {
            udoDeclaration = null;
            state = 2;
          }
        }
        break;
      }
    }

    if (!reprocessCurrentLine) {
      i++;
    }
  }

  /* HANDLE RESERVED GLOBAL VARIABLES */
  if (kr != null && ksmps == null && sr != null) {
    try {
      const krDouble = parseFloat(kr);
      const srDouble = parseFloat(sr);
      if (!isNaN(krDouble) && !isNaN(srDouble) && krDouble !== 0) {
        ksmps = Math.floor(srDouble / krDouble).toString();
      }
    } catch {
      ksmps = null;
    }
  }

  if (sr != null) {
    data.getProjectProperties().sampleRate = sr;
    data
      .getScore()
      .getTimeContext()
      .setSampleRate(parseInt(sr, 10) || 44100);
  }

  if (ksmps != null) {
    data.getProjectProperties().ksmps = ksmps;
  }

  data.getGlobalOrcSco().setGlobalOrc(globalOrch);
}

export function parseCsScore(data: BlueData, scoreText: string, importMode: CSDImportMode): void {
  let tables = '';
  let iStatements = '';

  const lines = scoreText.split('\n');

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();

    if (line.startsWith('f')) {
      tables += lines[i] + '\n';

      if (i < lines.length - 1) {
        do {
          const nextLine = lines[i + 1].trim();
          if (nextLine.length === 0) break;

          const c = nextLine.charAt(0);
          if ((c >= '0' && c <= '9') || c === '"' || c === '.') {
            tables += lines[i + 1] + '\n';
            i++;
          } else {
            break;
          }
        } while (i < lines.length - 1);
      }
    } else if (line.startsWith('i')) {
      iStatements += line + '\n';

      if (i < lines.length - 1) {
        do {
          const nextLine = lines[i + 1].trim();
          if (nextLine.length === 0) break;

          const c = nextLine.charAt(0);
          if ((c >= '0' && c <= '9') || c === '"' || c === '.') {
            iStatements += lines[i + 1] + '\n';
            i++;
          } else {
            break;
          }
        } while (i < lines.length - 1);
      }
    } else if (line.startsWith('s')) {
      iStatements += line + '\n';
    } else if (line.startsWith('t')) {
      if (line.length > 1) {
        const tLine = line.substring(1).trim();
        const tempoMap = data.getScore().getTimeContext().getTempoMap();
        const parts = tLine.split(/\s+/);

        if (parts.length % 2 === 0) {
          try {
            tempoMap.reset();
            for (let j = 0; j < parts.length; j += 2) {
              const beat = Number(parts[j]);
              const tempo = Number(parts[j + 1]);
              if (!Number.isFinite(beat) || !Number.isFinite(tempo)) {
                throw new Error('Invalid tempo statement found');
              }
              if (j === 0) {
                tempoMap.setTempoPoint(0, beat, tempo, CurveType.CONSTANT);
              } else {
                tempoMap.addTempoPoint(new TempoPoint(beat, tempo, CurveType.CONSTANT));
              }
            }
            tempoMap.setEnabled(true);
          } catch {
            throw new Error('Invalid tempo statement found');
          }
        } else {
          throw new Error('Invalid tempo statement found');
        }
      }
    }
  }

  const noteText = iStatements;

  const rootScore = data.getScore();
  if (rootScore.length > 0 && rootScore[0] instanceof PolyObject) {
    const pObj = rootScore[0] as PolyObject;
    if (pObj.length === 1 && pObj[0].length === 0) {
      pObj.length = 0;
    }
  }

  switch (importMode) {
    case CSDImportMode.IMPORT_GLOBAL: {
      if (rootScore.length > 0 && rootScore[0] instanceof PolyObject) {
        (rootScore[0] as PolyObject).newLayerAt(-1);
      }
      data.getGlobalOrcSco().setGlobalSco(noteText);
      break;
    }

    case CSDImportMode.IMPORT_SINGLE_SOUNDOBJECT: {
      const sections = getScoreSections(noteText);
      for (const section of sections) {
        setSoundObjectPerSection(data, section);
      }
      break;
    }

    case CSDImportMode.IMPORT_SOUNDOBJECT_PER_INSTRUMENT: {
      const sections = getScoreSections(noteText);
      for (const section of sections) {
        setSoundObjectsPerInstrument(data, section);
      }
      break;
    }
  }

  data.getTableSet().setTables(tables);
}

function isInstrOrUDODeclarationBoundary(trimmedLine: string): boolean {
  return (
    trimmedLine.startsWith('opcode') ||
    trimmedLine.startsWith('instr') ||
    trimmedLine.startsWith('endop') ||
    trimmedLine.startsWith('endin')
  );
}

function getScoreSections(scoreText: string): ScoreSection[] {
  const scoreSections: ScoreSection[] = [];
  const lines = scoreText.split('\n');

  let currentSection = '';
  let sectionStartTime = 0.0;

  for (const rawLine of lines) {
    const line = rawLine.trim();

    if (line.startsWith('s')) {
      const section: ScoreSection = {
        scoreText: currentSection,
        sectionStartTime,
      };
      scoreSections.push(section);

      const nl = getNotes(section.scoreText);
      sectionStartTime += getTotalDuration(nl);
      currentSection = '';
    } else {
      currentSection += rawLine + '\n';
    }
  }

  const section: ScoreSection = {
    scoreText: currentSection,
    sectionStartTime,
  };
  scoreSections.push(section);

  return scoreSections;
}

function setSoundObjectPerSection(data: BlueData, section: ScoreSection): void {
  const context = data.getScore().getTimeContext();
  const genScore = createSizedGenericScore(section.scoreText, 'Imported Score', context);
  genScore.setStartTime(TimePosition.beats(section.sectionStartTime));

  const rootScore = data.getScore();
  if (rootScore.length > 0 && rootScore[0] instanceof PolyObject) {
    const pObj = rootScore[0] as PolyObject;
    const sLayer = pObj.newLayerAt(-1);
    sLayer.push(genScore);
  }
}

function setSoundObjectsPerInstrument(data: BlueData, section: ScoreSection): void {
  const context = data.getScore().getTimeContext();
  const map = new Map<number, string[]>();

  const lines = section.scoreText.split('\n');
  let previousNote: Note | null = null;

  for (const line of lines) {
    if (!line.trim()) continue;
    const note = Note.createNoteFromText(line, previousNote);
    if (!note) continue;

    const iNumStr = note.getPField(1);
    if (!iNumStr) continue;

    const iNum = parseInt(iNumStr, 10);
    if (isNaN(iNum)) continue;

    if (!map.has(iNum)) {
      map.set(iNum, []);
    }
    map.get(iNum)!.push(line);
    previousNote = note;
  }

  const sortedKeys = Array.from(map.keys()).sort((a, b) => a - b);

  for (const iNum of sortedKeys) {
    const linesForInstr = map.get(iNum)!;
    const score = linesForInstr.join('\n') + '\n';
    const notes = getNotes(score);

    if (notes.length === 0) continue;

    notes.sort();
    const minStart = notes.getNote(0).getStartTime();

    notes.normalizeNoteList();

    const genScore = createSizedGenericScore(notes.toScoreText(), `Instrument ${iNum}`, context);

    genScore.setStartTime(TimePosition.beats(minStart + section.sectionStartTime));

    const rootScore = data.getScore();
    if (rootScore.length > 0 && rootScore[0] instanceof PolyObject) {
      const pObj = rootScore[0] as PolyObject;
      const sLayer = pObj.newLayerAt(-1);
      sLayer.push(genScore);
    }
  }
}

function createSizedGenericScore(
  noteText: string,
  name: string,
  context: TimeContext,
): GenericScore {
  const genScore = new GenericScore();
  genScore.setScoreText(noteText);
  const notes = getNotes(noteText);
  const totalDur = getTotalDuration(notes);
  genScore.setSubjectiveDuration(TimeDuration.beats(totalDur));
  genScore.setName(name);
  return genScore;
}
